Core/Src/steeringio.c

```diff
@@ -0,0 +1,107 @@
+#include "steeringio.h"
+#include <string.h>
+#include <assert.h>
+#include <stdlib.h>
+#include "cmsis_os.h"
+#include "cerberus_conf.h"
+
+static osMutexAttr_t steeringio_data_mutex_attributes;
+static osMutexAttr_t steeringio_ringbuffer_mutex_attributes;
+
+static void debounce_cb(void const *arg);
+osTimerDef(debounce_timer, debounce_cb);
+
+enum {
+	NOT_PRESSED,
+	PRESSED
+};
+
+/* Creates a new Steering Wheel interface */
+steeringio_t *steeringio_init()
+{
+	steeringio_t *steeringio = malloc(sizeof(steeringio_t));
+	assert(steeringio);
+
+	/* Create Mutexes */
+	steeringio->ringbuffer_mutex = osMutexNew(&steeringio_data_mutex_attributes);
+	assert(steeringio->ringbuffer_mutex);
+
+	steeringio->button_mutex = osMutexNew(&steeringio_ringbuffer_mutex_attributes);
+	assert(steeringio->button_mutex);
+
+	//TODO: Error handler for allocation
+	for (uint8_t i = 0; i < MAX_STEERING_BUTTONS; i++) {
+		steeringio->debounce_timers[i] = osTimerCreate(osTimer(debounce_timer), osTimerOnce, steeringio);
+		assert(steeringio->debounce_timers[i]);
+	}
+
+	steeringio->debounce_buffer = ringbuffer_create(MAX_STEERING_BUTTONS, sizeof(uint8_t));
+	assert(steeringio->debounce_buffer);
+
+	return steeringio;
+}
+
+bool get_steeringio_button(steeringio_t *wheel, steeringio_button_t button)
+{
+	if (!wheel)
+		return 0;
+
+	bool ret = 0;
+
+	osMutexAcquire(wheel->button_mutex, osWaitForever);
+	ret = wheel->debounced_buttons[button];
+	osMutexRelease(wheel->button_mutex);
+
+	return ret;
+}
+
+/* Callback to see if we have successfully debounced */
+static void debounce_cb(void const *arg)
+{
+	steeringio_t *wheel = (steeringio_t *)arg;
+	if (!wheel)
+		return;
+
+	steeringio_button_t button;
+
+	/*
+	 * Pop off most recent button being debounced
+	 * Note: timers are started sequentially, therefore buttons
+	 *  will be popped off in order
+	 *
+	 * Also needs to be atomic
+	 */
+	osMutexAcquire(wheel->ringbuffer_mutex, osWaitForever);
+	ringbuffer_dequeue(wheel->debounce_buffer, &button);
```

Comment:
this isnt really specfic to here but if were gonna start using asserts we need to be extra thoughtful about there interaction with mutexes. Bad enough to get stuck in a dummy handler, having threads locked because of it seems like another layer of headaches